src/securemul.rs

```diff
@@ -0,0 +1,430 @@
+        let DValue {
+            d: left_d,
+            index: left_index,
+        } = ctx.helper_ring.receive(HelperAddr::Left).await?;
+
+        // sanity check to make sure they've computed it using the same seed
+        if left_index == self.index {
+            // now we are ready to construct the result - 2/3 secret shares of a * b.
+            let lhs = a0 * b0 + F::from(left_d) + s0;
+            let rhs = a1 * b1 + s1 + F::from(right_d);
+
+            Ok(ReplicatedSecretSharing::new(lhs, rhs))
+        } else {
+            Err(Box::new(Error::IndexMismatch {
+                my_index: self.index,
+                their_index: left_index,
+            }))
+        }
+    }
+}
+
+/// Module to support streaming interface for secure multiplication
+pub mod stream {
+    use crate::error::BoxError;
+    use crate::field::Field;
+    use crate::helpers::ring::Ring;
+    use crate::replicated_secret_sharing::ReplicatedSecretSharing;
+    use crate::securemul::{ProtocolContext, SecureMul};
+    use futures::{ready, Stream};
+    use pin_project::pin_project;
+    use std::future::Future;
+    use std::mem;
+    use std::pin::Pin;
+    use std::task::{Context, Poll};
+    use tracing::error;
+
+    /// A variant of stream transform that combines semantic of `StreamExt::chunks` and `StreamExt::scan`.
+    /// Consumes the input stream and keeps accumulating items in the internal buffer until it reaches
+    /// `capacity` elements. Then the elements are moved to the `f` function that must produce a future
+    /// resolvable to the same type as element type of the input stream.
+    ///
+    /// When elements are given to the `f` function, no other elements will be taken off the input stream
+    /// until the future returned by it is resolved. It is important to note that the resulting item
+    /// returned by this function is kept in the buffer, so next time stream is polled, only (`capacity`-1)
+    /// elements will be polled off before calling `f` again.
+    ///
+    /// If input stream yields `None` while buf does not have at least `capacity` elements, `f` will
+    /// be called on partial buf
+    #[pin_project]
+    pub struct ChunkScan<St: Stream, F, Fut> {
+        /// Input stream
+        #[pin]
+        stream: St,
+
+        /// how many elements to keep in the buffer before calling `f`
+        capacity: usize,
+
+        /// Buffer for items taken off the input stream
+        buf: Vec<St::Item>,
+
+        /// Transforms Vec<Item> -> Future<Output=Result<Item, Error>>
+        f: F,
+
+        /// future in progress
+        #[pin]
+        future: Option<Fut>,
+    }
+
+    impl<St, F, Fut> Stream for ChunkScan<St, F, Fut>
+    where
+        St: Stream,
+        St::Item: Clone,
+        F: FnMut(Vec<St::Item>) -> Fut,
+        Fut: Future<Output = Result<St::Item, BoxError>>,
+    {
+        type Item = St::Item;
+
+        fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
```

Comment:
I'm surprised to see this living inside of securemul.rs - it seems generic and not scoped to just secure multiplication.